hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/batch/ByteBufferMessageBatchFactory.java

```diff
@@ -0,0 +1,47 @@
+package pl.allegro.tech.hermes.consumers.consumer.batch;
+
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.common.exception.InternalProcessingException;
+import pl.allegro.tech.hermes.common.metric.Gauges;
+import pl.allegro.tech.hermes.common.metric.HermesMetrics;
+
+import java.nio.ByteBuffer;
+import java.time.Clock;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static java.lang.String.format;
+import static java.util.UUID.randomUUID;
+
+public class ByteBufferMessageBatchFactory implements MessageBatchFactory {
+    private final DirectBufferPool bufferPool;
+    private final Clock clock;
+
+    public ByteBufferMessageBatchFactory(int poolableSize, int maxPoolSize, Clock clock, HermesMetrics hermesMetrics) {
+        this.clock = clock;
+        this.bufferPool = new DirectBufferPool(maxPoolSize, poolableSize, true);
+        hermesMetrics.registerGauge(Gauges.BATCH_BUFFER_TOTAL_BYTES, bufferPool::totalMemory);
+        hermesMetrics.registerGauge(Gauges.BATCH_BUFFER_AVAILABLE_BYTES, bufferPool::availableMemory);
+    }
+
+    @Override
+    public MessageBatch createBatch(Subscription subscription) {
+        try {
+            ByteBuffer buffer = bufferPool.allocate(subscription.getBatchSubscriptionPolicy().getBatchVolume());
+            switch (subscription.getContentType()) {
+                case JSON:
+                    return new JsonMessageBatch(randomUUID().toString(), buffer, subscription, clock);
+                case AVRO:
```

Comment:
We need to do some research to check whether there is a better solution other than sending message with a batch schema that contains list of events with message schema